Mathlib/Data/Int/GCD.lean

```diff
@@ -511,255 +513,270 @@ theorem pow_gcd_eq_one {M : Type _} [Monoid M] (x : M) {m n : ℕ} (hm : x ^ m =
+
+theorem isNat_not_coprime : {x y nx ny : ℕ} →
+    IsNat x nx → IsNat y ny → ¬ Nat.coprime nx ny → ¬ Nat.coprime x y
+  | _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, h => h
+
+theorem isInt_gcd : {x y nx ny : ℤ} → {z : ℕ} →
+    IsInt x nx → IsInt y ny → Int.gcd nx ny = z → IsNat (Int.gcd x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+theorem isInt_lcm : {x y nx ny : ℤ} → {z : ℕ} →
+    IsInt x nx → IsInt y ny → Int.lcm nx ny = z → IsNat (Int.lcm x y) z
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨rfl⟩
+
+/-- Supporting definition for `proveNatGCD`. Returns the GCD and an equality proof. -/
+def proveNatGCD' (x y : ℕ) : ((d : ℕ) × Q(Nat.gcd $x $y = $d)) :=
+  match x, y with
+  | 0, y => ⟨y, q(Nat.gcd_zero_left $y)⟩
+  | x, 0 => ⟨x, q(Nat.gcd_zero_right $x)⟩
+  | 1, y => ⟨1, q(Nat.gcd_one_left $y)⟩
+  | x, 1 => ⟨1, q(Nat.gcd_one_right $x)⟩
+  | x, y =>
+    let (d, a, b) := Nat.xgcdAux x 1 0 y 0 1
+    if d = x then
+      have q : ℕ := y / x
+      have pq : Q(Nat.mul $x $q = $y) := (q(Eq.refl $y) : Expr)
+      ⟨x, q(nat_gcd_helper_dvd_left $x $y $q $pq)⟩
+    else if d = y then
+      have q : ℕ := x / y
+      have pq : Q(Nat.mul $y $q = $x) := (q(Eq.refl $x) : Expr)
+      ⟨y, q(nat_gcd_helper_dvd_right $x $y $q $pq)⟩
+    else
+      have a' : ℕ := a.natAbs
+      have b' : ℕ := b.natAbs
+      have u : ℕ := x / d
+      have v : ℕ := y / d
+      have pu : Q(Nat.mul $d $u = $x) := (q(Eq.refl $x) : Expr)
+      have pv : Q(Nat.mul $d $v = $y) := (q(Eq.refl $y) : Expr)
+      have tx : ℕ := x * a'
+      have ty : ℕ := y * b'
+      have px : Q(Nat.mul $x $a' = $tx) := (q(Eq.refl $tx) : Expr)
+      have py : Q(Nat.mul $y $b' = $ty) := (q(Eq.refl $ty) : Expr)
+      if a ≥ 0 then
+        have pt : Q(Nat.add $ty $d = $tx) := (q(Eq.refl $tx) : Expr)
+        ⟨d, q(nat_gcd_helper_2 $d $x $y $a' $b' $u $v $tx $ty $pu $pv $px $py $pt)⟩
+      else
+        have pt : Q(Nat.add $tx $d = $ty) := (q(Eq.refl $ty) : Expr)
+        ⟨d, q(nat_gcd_helper_1 $d $x $y $a' $b' $u $v $tx $ty $pu $pv $px $py $pt)⟩
+
+/-- Given natural number literals `nx` and `ny`, return their GCD as a natural number literal
+and an equality proof. -/
+def proveNatGCD (nx ny : Q(ℕ)) : Option ((d : Q(ℕ)) × Q(Nat.gcd $nx $ny = $d)) := do
+  let x ← nx.natLit?
+  let y ← ny.natLit?
+  let ⟨c, pf⟩ := proveNatGCD' x y
+  return ⟨mkRawNatLit c, pf⟩
+#align tactic.norm_num.prove_gcd_nat Tactic.NormNum.proveNatGCD
+
+@[norm_num Nat.gcd _ _]
+def evalNatGCD : NormNumExt where eval {u α} e := do
+  let .app (.app _ (x : Q(ℕ))) (y : Q(ℕ)) ← Meta.whnfR e | failure
+  let sℕ : Q(AddMonoidWithOne ℕ) := q(instAddMonoidWithOneNat)
+  let ⟨nx, p⟩ ← deriveNat x sℕ
+  let ⟨ny, q⟩ ← deriveNat y sℕ
+  let ⟨cd, pf⟩ ← proveNatGCD nx ny
+  let pf' : Q(IsNat (Nat.gcd $x $y) $cd) := q(isNat_gcd $p $q $pf)
+  return .isNat sℕ cd pf'
+
+/-- Supporting definition for `proveNatLCM`. Returns the LCM and an equality proof. -/
+def proveNatLCM' (x y : ℕ) : ((d : ℕ) × Q(Nat.lcm $x $y = $d)) :=
```

Comment:
```suggestion
def proveNatLCM' (x y : Q(ℕ)) : ((d : ℕ) × Q(Nat.lcm $x $y = $d)) :=
```
You should retain the original natLit expressions, so that they can be embedded in the generated expressions without having to recompute the hash and breaking pointer equality for those subterms. I think you should never need to use `natLit?`: either you know that the input expr is a natlit in which case you can use `.natLit!`, or you don't and you want to use `deriveNat x` instead.

Reply:
It's now sharing as much as possible, except for some 0's and 1's here and there.

I originally used `natLit?` because partial functions that use `panic!` make me a little uncomfortable, but I switched it over to use `natLit!` anyway.